Plugin doesn't support reproducible build

Hi,

Given a build with 

```xml
<properties>
....
    <project.build.outputTimestamp>2025-01-01T00:00:00Z</project.build.outputTimestamp>
...
</properties>
```

and spdx plugin

```xml
        <plugin>
          <groupId>org.spdx</groupId>
          <artifactId>spdx-maven-plugin</artifactId>
          <version>1.0.0-RC1</version>
          <executions>
            <execution>
              <id>build-spdx</id>
              <goals>
                <goal>createSPDX</goal>
              </goals>
            </execution>
          </executions>
        </plugin>
```

Then

```
mvn clean install
mvn verify artifact:compare
```

```
[ERROR] sha512 mismatch **.spdx.json: investigate with diffoscope ****.spdx.json target/site/l*****.spdx.json
[ERROR] Reproducible Build output summary: 3 files ok, 1 different
```

Using diffoscope like recommended by maven show difference because of the date on the spdx license

```diff
 {
      "SPDXID": "SPDXRef-DOCUMENT",
      "creationInfo": {
-        "created": "2025-02-19T07:46:56Z",
+        "created": "2025-02-19T07:48:32Z",
         "creators": [
             "Tool: spdx-maven-plugin"
         ],
         "licenseListVersion": "3.26.0"
     },
     "dataLicense": "CC0-1.0",
```

--- a/src/it/advanced/pom.xml
+++ b/src/it/advanced/pom.xml
@@ -41,6 +41,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <maven.compiler.release>11</maven.compiler.release>
         <jena.version>3.10.0</jena.version>
+        <project.build.outputTimestamp>1740169111</project.build.outputTimestamp>
     </properties>
 
     <dependencies>

--- a/src/main/java/org/spdx/maven/CreateSpdxMojo.java
+++ b/src/main/java/org/spdx/maven/CreateSpdxMojo.java
@@ -40,6 +40,7 @@ import org.apache.maven.shared.model.fileset.FileSet;
 import org.spdx.core.InvalidSPDXAnalysisException;
 import org.spdx.core.SpdxCoreConstants.SpdxMajorVersion;
 import org.spdx.library.SpdxModelFactory;
+import org.spdx.library.model.v2.SpdxConstantsCompatV2;
 import org.spdx.maven.utils.LicenseMapperException;
 import org.spdx.maven.utils.SpdxBuilderException;
 import org.spdx.maven.utils.SpdxCollectionException;
@@ -60,13 +61,16 @@ import org.apache.maven.artifact.resolver.filter.CumulativeScopeArtifactFilter;
 import java.io.File;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.text.SimpleDateFormat;
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
 import java.util.*;
 import java.util.Map.Entry;
 
 /**
- * NOTE: Currently this is a prototype plugin for supporting SPDX in a Maven build.
+ * NOTE: Plugin for supporting SPDX in a Maven build.
  * <p>
- * Goal which creates a new SPDX file for the package being built.  Will replace any existing SPDX file.
+ * Goal which creates new SPDX file(s) for the package being built.  Will replace any existing SPDX file(s).
  * <p>
  * All SPDX document and SPDX package properties are supported as parameters to the plugin.
  * <p>
@@ -86,6 +90,7 @@ import java.util.Map.Entry;
  * <li>package supplier: project organization</li>
  * <li>package versionInfo: project version</li>
  * <li>files for analysis: build source files + project resource files</li>
+ * <li>created: creation time of the SPDX document</li>
  * </ul><p>
  * Additional SPDX fields are supplied as configuration parameters to this plugin.
  */
@@ -530,6 +535,17 @@ public class CreateSpdxMojo extends AbstractMojo
      */
     @Parameter( defaultValue = "true" )
     private boolean includeCompileScope;
+
+    /**
+     * SPDX Creation timestamp either formatted as ISO 8601
+     * <code>yyyy-MM-dd'T'HH:mm:ssXXX</code> or as an int representing seconds since the epoch.
+     *
+     * Reference: Reproducible <a href="https://reproducible-builds.org/docs/source-date-epoch/">Code Source-Date-Epoch spec</a>
+     *
+     * @since 1.0.0
+     */
+    @Parameter( defaultValue = "${project.build.outputTimestamp}" )
+    private String created;
 
     public void execute() throws MojoExecutionException
     {
@@ -978,6 +994,36 @@ public class CreateSpdxMojo extends AbstractMojo
         
         final Packaging packaging = Packaging.valueOfPackaging(  mavenProject.getPackaging() );
         retval.setPackaging( packaging != null ? packaging : Packaging.JAR );
+        Date createdDate;
+        getLog().info( "Created: " + this.created );
+        if ( this.created == null || this.created.isEmpty() )
+        {
+            createdDate = new Date();
+        }
+        else
+        {
+            // The created parameter can be either a seconds from epoch or an ISO 8601 format
+            // We'll check for the int value first
+            if ( created.matches( "\\d+" ) )
+            {
+                createdDate = new Date( Integer.parseInt( created ) );
+            }
+            else
+            {
+                try
+                {
+                    createdDate = Date.from( Instant.parse( created ) );
+                }
+                catch ( DateTimeParseException ex)
+                {
+                    getLog().warn( "Invalid created date " + created + ".  Using current date-time" );
+                    createdDate = new Date();
+                }
+
+            }
+
+        }
+        retval.setCreated( new SimpleDateFormat( SpdxConstantsCompatV2.SPDX_DATE_FORMAT).format( createdDate ) );
         return retval;
     }
 

--- a/src/main/java/org/spdx/maven/CreateSpdxMojo.java
+++ b/src/main/java/org/spdx/maven/CreateSpdxMojo.java
@@ -1026,7 +1026,9 @@ public class CreateSpdxMojo extends AbstractMojo
             }
 
         }
-        retval.setCreated( new SimpleDateFormat( SpdxConstantsCompatV2.SPDX_DATE_FORMAT).format( createdDate ) );
+        SimpleDateFormat dateFormat = new SimpleDateFormat( SpdxConstantsCompatV2.SPDX_DATE_FORMAT);
+        dateFormat.setTimeZone( TimeZone.getTimeZone("GMT") );
+        retval.setCreated( dateFormat.format( createdDate ) );
         return retval;
     }
 

--- a/src/main/java/org/spdx/maven/CreateSpdxMojo.java
+++ b/src/main/java/org/spdx/maven/CreateSpdxMojo.java
@@ -951,7 +951,7 @@ public class CreateSpdxMojo extends AbstractMojo
         if ( mainArtifact != null && mainArtifact.getFile() != null )
         {
             packageFileName = mainArtifact.getArtifactId() + "-" + mainArtifact.getVersion() + "." + mainArtifact.getType();
-            packageFile = new File(mainArtifact.getFile().getParent() + File.separator + packageFileName);
+            packageFile = new File( mainArtifact.getFile().getParent() + File.separator + packageFileName );
         }
 
         Set<Checksum> checksums = null;
@@ -959,7 +959,7 @@ public class CreateSpdxMojo extends AbstractMojo
         {
             try
             {
-                getLog().debug( "Generating checksum for file "+packageFile.getAbsolutePath() );
+                getLog().debug( "Generating checksum for file " + packageFile.getAbsolutePath() );
                 Set<String> algorithms = getChecksumAlgorithms();
                 checksums = AbstractFileCollector.generateChecksum( packageFile, algorithms );
             }
@@ -991,11 +991,14 @@ public class CreateSpdxMojo extends AbstractMojo
         retval.setDocumentAnnotations( this.documentAnnotations );
         retval.setPackageAnnotations( this.packageAnnotations );
         retval.setExternalRefs( this.externalReferences );
-        
-        final Packaging packaging = Packaging.valueOfPackaging(  mavenProject.getPackaging() );
+
+        final Packaging packaging = Packaging.valueOfPackaging( mavenProject.getPackaging() );
         retval.setPackaging( packaging != null ? packaging : Packaging.JAR );
         Date createdDate;
-        getLog().info( "Created: " + this.created );
+        if ( this.created == null || this.created.isEmpty() )
+        {
+            this.created = mavenProject.getModel().getProperties().getProperty( "project.build.outputTimestamp" );
+        }
         if ( this.created == null || this.created.isEmpty() )
         {
             createdDate = new Date();
@@ -1006,7 +1009,7 @@ public class CreateSpdxMojo extends AbstractMojo
             // We'll check for the int value first
             if ( created.matches( "\\d+" ) )
             {
-                createdDate = new Date( Integer.parseInt( created ) );
+                createdDate = new Date( Long.parseLong( created ) * 1000 );
             }
             else
             {

--- a/src/main/java/org/spdx/maven/utils/SpdxProjectInformation.java
+++ b/src/main/java/org/spdx/maven/utils/SpdxProjectInformation.java
@@ -55,6 +55,7 @@ public class SpdxProjectInformation
     private ExternalReference[] externalRefs;
     private Set<Checksum> checksums;
     private Packaging packaging;
+    private String created;
     
     
 
@@ -353,6 +354,7 @@ public class SpdxProjectInformation
         LOG.debug( "SPDX Declared license: {}", this.getDeclaredLicense() );
         LOG.debug( "SPDX Download URL: {}", this.getDownloadUrl() );
         LOG.debug( "SPDX Home page: {}", this.getHomePage() );
+        LOG.debug( "SPDX Created: {}", this.getCreated() );
         if ( this.documentAnnotations != null && this.documentAnnotations.length > 0 )
         {
             LOG.debug( "Document annotations: " );
@@ -443,4 +445,8 @@ public class SpdxProjectInformation
     {
         return this.externalRefs;
     }
+
+    public void setCreated( String created ) { this.created = created; }
+
+    public String getCreated() { return this.created; }
 }

--- a/src/main/java/org/spdx/maven/utils/SpdxV2DocumentBuilder.java
+++ b/src/main/java/org/spdx/maven/utils/SpdxV2DocumentBuilder.java
@@ -10,12 +10,7 @@ import java.io.FileOutputStream;
 import java.io.IOException;
 import java.net.URI;
 import java.security.NoSuchAlgorithmException;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Date;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
 
 import org.apache.maven.model.License;
 import org.apache.maven.plugin.MojoExecutionException;
@@ -123,6 +118,8 @@ public class SpdxV2DocumentBuilder
             {
                 spdxDoc.setComment( projectInformation.getDocumentComment() );
             }
+            // created
+            Objects.requireNonNull( spdxDoc.getCreationInfo() ).setCreated( projectInformation.getCreated() );
             // creator
             fillCreatorInfo( projectInformation );
             // data license

--- a/src/main/java/org/spdx/maven/utils/SpdxV2DocumentBuilder.java
+++ b/src/main/java/org/spdx/maven/utils/SpdxV2DocumentBuilder.java
@@ -118,8 +118,6 @@ public class SpdxV2DocumentBuilder
             {
                 spdxDoc.setComment( projectInformation.getDocumentComment() );
             }
-            // created
-            Objects.requireNonNull( spdxDoc.getCreationInfo() ).setCreated( projectInformation.getCreated() );
             // creator
             fillCreatorInfo( projectInformation );
             // data license
@@ -194,7 +192,7 @@ public class SpdxV2DocumentBuilder
                 LOG.warn( "Invalid creator string ( {} ), {} will be skipped.", verify, parameterCreator );
             }
         }
-        SpdxCreatorInformation spdxCreator = spdxDoc.createCreationInfo( creators, format.format( new Date() ) );
+        SpdxCreatorInformation spdxCreator = spdxDoc.createCreationInfo( creators, projectInformation.getCreated() );
         spdxCreator.setComment( projectInformation.getCreatorComment() );
         spdxCreator.setLicenseListVersion( ListedLicenses.getListedLicenses().getLicenseListVersion() );
         spdxDoc.setCreationInfo( spdxCreator );

--- a/src/main/java/org/spdx/maven/utils/SpdxV3DocumentBuilder.java
+++ b/src/main/java/org/spdx/maven/utils/SpdxV3DocumentBuilder.java
@@ -137,6 +137,8 @@ public class SpdxV3DocumentBuilder
             {
                 spdxDoc.setComment( projectInformation.getDocumentComment() );
             }
+            // created
+            requireNonNull( spdxDoc.getCreationInfo() ).setCreated( projectInformation.getCreated() );
             // creator
             fillCreatorInfo( projectInformation );
             // data license

--- a/src/main/java/org/spdx/maven/utils/SpdxV3DocumentBuilder.java
+++ b/src/main/java/org/spdx/maven/utils/SpdxV3DocumentBuilder.java
@@ -137,8 +137,6 @@ public class SpdxV3DocumentBuilder
             {
                 spdxDoc.setComment( projectInformation.getDocumentComment() );
             }
-            // created
-            requireNonNull( spdxDoc.getCreationInfo() ).setCreated( projectInformation.getCreated() );
             // creator
             fillCreatorInfo( projectInformation );
             // data license
@@ -216,6 +214,7 @@ public class SpdxV3DocumentBuilder
     {
         CreationInfo creationInfo = spdxDoc.getCreationInfo();
         requireNonNull( creationInfo, "CreationInfo for the SPDX document must not be null" );
+        creationInfo.setCreated( projectInformation.getCreated() );
         String[] parameterCreators = projectInformation.getCreators();
         for ( String parameterCreator : parameterCreators )
         {

--- a/src/test/java/org/spdx/maven/TestSpdxV2Mojo.java
+++ b/src/test/java/org/spdx/maven/TestSpdxV2Mojo.java
@@ -189,7 +189,7 @@ public class TestSpdxV2Mojo extends AbstractMojoTestCase
         assertTrue( foundCreator1 );
         assertTrue( foundCreator2 );
         // created
-        assertEquals( "2025-02-21T20:18:31Z", result.getCreationInfo().getCreated() );
+        assertEquals( "2025-02-21T12:18:31Z", result.getCreationInfo().getCreated() );
         // package parameters
         assertEquals( 1, result.getDocumentDescribes().size() );
         SpdxElement described = result.getDocumentDescribes().toArray( new SpdxElement[result.getDocumentDescribes().size()] )[0];

--- a/src/test/java/org/spdx/maven/TestSpdxV2Mojo.java
+++ b/src/test/java/org/spdx/maven/TestSpdxV2Mojo.java
@@ -11,6 +11,9 @@ import java.util.Collections;
 import java.util.List;
 
 import org.apache.maven.plugin.testing.AbstractMojoTestCase;
+import org.codehaus.plexus.util.ReaderFactory;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+import org.codehaus.plexus.util.xml.Xpp3DomBuilder;
 import org.junit.After;
 import org.junit.AfterClass;
 import org.junit.Before;
@@ -71,6 +74,8 @@ public class TestSpdxV2Mojo extends AbstractMojoTestCase
     public void testExecute() throws Exception
     {
         File testPom = new File( getBasedir(), UNIT_TEST_RESOURCE_DIR + "/pom.xml" );
+        Xpp3Dom pluginPomDom = Xpp3DomBuilder.build(ReaderFactory.newXmlReader(testPom));
+        System.out.println("Created time test: "+ pluginPomDom.getChild( "properties" ).getChild( "project.build.outputTimestamp" ));
         //        CreateSpdxMojo mojo = (CreateSpdxMojo) configureMojo( myMojo, "spdx-maven-plugin", testPom );
         // if the below does not work due to a lookup error, run mvn test goal
         CreateSpdxMojo mojo = (CreateSpdxMojo) lookupMojo( "createSPDX", testPom );
@@ -183,6 +188,8 @@ public class TestSpdxV2Mojo extends AbstractMojoTestCase
         }
         assertTrue( foundCreator1 );
         assertTrue( foundCreator2 );
+        // created
+        assertEquals( "2025-02-21T20:18:31Z", result.getCreationInfo().getCreated() );
         // package parameters
         assertEquals( 1, result.getDocumentDescribes().size() );
         SpdxElement described = result.getDocumentDescribes().toArray( new SpdxElement[result.getDocumentDescribes().size()] )[0];

--- a/src/test/java/org/spdx/maven/TestSpdxV2Mojo.java
+++ b/src/test/java/org/spdx/maven/TestSpdxV2Mojo.java
@@ -189,7 +189,7 @@ public class TestSpdxV2Mojo extends AbstractMojoTestCase
         assertTrue( foundCreator1 );
         assertTrue( foundCreator2 );
         // created
-        assertEquals( "2025-02-21T12:18:31Z", result.getCreationInfo().getCreated() );
+        assertEquals( "2025-02-21T20:18:31Z", result.getCreationInfo().getCreated() );
         // package parameters
         assertEquals( 1, result.getDocumentDescribes().size() );
         SpdxElement described = result.getDocumentDescribes().toArray( new SpdxElement[result.getDocumentDescribes().size()] )[0];

--- a/src/test/java/org/spdx/maven/TestSpdxV3Mojo.java
+++ b/src/test/java/org/spdx/maven/TestSpdxV3Mojo.java
@@ -195,7 +195,7 @@ public class TestSpdxV3Mojo extends AbstractMojoTestCase
         }
         assertTrue( foundCreator1 );
         assertTrue( foundCreator2 );
-        assertEquals( "2025-02-21T12:18:31Z", result.getCreationInfo().getCreated() );
+        assertEquals( "2025-02-21T20:18:31Z", result.getCreationInfo().getCreated() );
         assertEquals( 1, result.getCreationInfo().getCreatedUsings().size() );
         assertEquals( "spdx-maven-plugin", result.getCreationInfo().getCreatedUsings().iterator().next().getName().get() );
         // package parameters

--- a/src/test/java/org/spdx/maven/TestSpdxV3Mojo.java
+++ b/src/test/java/org/spdx/maven/TestSpdxV3Mojo.java
@@ -195,6 +195,7 @@ public class TestSpdxV3Mojo extends AbstractMojoTestCase
         }
         assertTrue( foundCreator1 );
         assertTrue( foundCreator2 );
+        assertEquals( "2025-02-21T12:18:31Z", result.getCreationInfo().getCreated() );
         assertEquals( 1, result.getCreationInfo().getCreatedUsings().size() );
         assertEquals( "spdx-maven-plugin", result.getCreationInfo().getCreatedUsings().iterator().next().getName().get() );
         // package parameters

--- a/src/test/resources/unit/spdx-maven-plugin-test/pom-v3.xml
+++ b/src/test/resources/unit/spdx-maven-plugin-test/pom-v3.xml
@@ -22,6 +22,7 @@
     </organization>
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <project.build.outputTimestamp>2025-02-21'T'12:18:31Z</project.build.outputTimestamp>
     </properties>
 
   <dependencies>

--- a/src/test/resources/unit/spdx-maven-plugin-test/pom.xml
+++ b/src/test/resources/unit/spdx-maven-plugin-test/pom.xml
@@ -22,6 +22,7 @@
     </organization>
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <project.build.outputTimestamp>1740169111</project.build.outputTimestamp>
     </properties>
 
   <dependencies>

